### ISSUE TITLE
Allow azure-cli to trust on OS certificates

### ISSUE THREAD

**Issue by joaocc** [Azure/azure-cli#28050]
**Related command**
any command (prob REQUESTS_CA_BUNDLE is the closest)

**Is your feature request related to a problem? Please describe.**
Yes. Enabling azure-cli to work with custom trusted cert roots.

On macOS it is not simple to configure https://learn.microsoft.com/en-us/cli/azure/use-cli-effectively?tabs=bash%2Cbash2#work-behind-a-proxy as:
1. the ca_bundle is located in a folder that depends on 2 variables: client version and python version (`/usr/local/Cellar/azure-cli/<cliversion>/libexec/lib/python<version>/site-packages/certifi/cacert.pem`)
1. every time we upgrade the client, the bundle has to be recreated

**Describe the solution you'd like**
Allow azure-cli to trust the certificates that are trusted by the OS (keychain on macOS, others on other OS).
This could be the default behaviour or, in order to keep compat with current behaviour, allow enabling it via both env var (such as `CA_BUNDLE_TRUST_OS`) and CLI option (`--ca-bundle-trust-os`)

Alternatively (or in conjuntcion), allow specification of an additional bundle vi both env var (such as `REQUESTS_CA_BUNDLE_EXTRA`) and CLI option (`--requests-ca-bundle-extra`)

**Describe alternatives you've considered**
Implementing bash scripts that need to try to guess the correct location to place new certs, generate a new bundle, and then rewriting.

**Additional context**

**Comment by yonzhan** [comment]
Thank you for opening this issue, we will look into it.

**Comment by codepic** [comment]
Looking through the source code of `azure-cli`, it seems like you have already 2 options:

1. Use the `AZURE_CLI_DISABLE_CONNECTION_VERIFICATION` (not recommended)
2. Use the `REQUESTS_CA_BUNDLE`

For the latter, you should be able to [export all CA certificates from Apple KeyChain into a single PEM bundle](https://stackoverflow.com/questions/41851601/how-do-i-export-all-ca-certificates-from-apple-keychain-into-a-single-pem-bundle)

**Comment by joaocc** [comment]
HI @codepic. Thanks for the update.
In our case, the reason for using custom certs is different - we are using a corp firewall that does traffic inspection, so we need to deploy the cert to all machines. As you can understand, this is not a matter of short-sighted cost savings.
In a more general note, any tool that doesn't use the OS certs requires additional effort (to plan, test, deploy, maintain) instead of trusting the user/developer/architect/IT staff to decide what is more suitable for their particular environment. Even Firefox ended up implementing a way to enable users to decide whether or not to trust the certs that the OS trusts (Edge and Chrome do that by default, as do many other tools and apps).
For the reasons above, the "trust OS certificates" option (or default) would be a much cleaner and simpler solution, and one that would not have any downside for existing users, but would facilitate the lives of many others.
I agree that the specific mechanics to enable it would be best determined by the team (as I was just providing illustrative examples).

**Comment by jiasli** [comment]
@joaocc, I fully understand your concern. @codepic, thank you very much for the detailed explanation.

The fact is that, Azure CLI internally depends on `requests` library and has no special logic for choosing and using CA bundle. There is an issue opened in `requests` repo for this feature request:

- https://github.com/psf/requests/issues/2966

**Comment by joaocc** [comment]
@jiasli, looking at the issue you pointed, it seems there is now what seems to be a simple(r) way to address this without waiting for requests (sorry if I am oversimplifying :)), but using `truststore`.

* https://truststore.readthedocs.io/en/latest/#using-truststore-with-requests

```
# Using truststore with Requests
# Just like with urllib3 using truststore.inject_into_ssl() is the easiest method for using Truststore with Requests:

import requests
import truststore

truststore.inject_into_ssl()

resp = requests.request("GET", "https://example.com")
```

I understand it might not be so simple to make a change in a core are as this, but... Worth trying to ask for this to be considered, especially due to the upside for the community of azure-cli users.
Thanks

**Comment by Ousret** [comment]
There's an another option. In fact two options.

- The first would be to replace requests by (drop-in replacement) [niquests](https://github.com/jawah/niquests), with implemented support for https://github.com/psf/requests/issues/2966 (amongst other things)
- Finally, you could consider using the wassima library to extract the system CA bundle automatically.

The caveat of truststore is the monkeypatching of the stdlib and that it is exclusive to Python 3.10+

**Comment by bodnarbm-gov** [comment]
> [@joaocc](https://github.com/joaocc), I fully understand your concern. [@codepic](https://github.com/codepic), thank you very much for the detailed explanation.
> 
> The fact is that, Azure CLI internally depends on `requests` library and has no special logic for choosing and using CA bundle. There is an issue opened in `requests` repo for this feature request:
> 
> * [Consider using system trust stores by default in 3.0.0. psf/requests#2966](https://github.com/psf/requests/issues/2966)

The referenced issue with requests is now closed as completed, with the resolution being the recommendation to use `truststore` to patch `requests`.

@jiasli could we evaluate conditionally adding/importing truststore. For most systems on python 3.10+, this should solve the issue with a better user experience then messing with the bundled certs. For systems where truststore does not work (python 3.9 or old versions of macos),  then the conditional import should not affect preset behavior.

**Comment by bodnarbm** [comment]
> [@joaocc](https://github.com/joaocc), I fully understand your concern. [@codepic](https://github.com/codepic), thank you very much for the detailed explanation.
> 
> The fact is that, Azure CLI internally depends on `requests` library and has no special logic for choosing and using CA bundle. There is an issue opened in `requests` repo for this feature request:
> 
> * [Consider using system trust stores by default in 3.0.0. psf/requests#2966](https://github.com/psf/requests/issues/2966)

The referenced issue with requests is now closed as completed, with the resolution being the recommendation to use `truststore` to patch `requests`.

@jiasli could we evaluate conditionally adding/importing truststore. For most systems on python 3.10+, this should solve the issue with a better user experience then messing with the bundled certs. For systems where truststore does not work (python 3.9 or old versions of macos),  then the conditional import should not affect preset behavior.

**Comment by StevenDevooght** [comment]
Will this be revisited? I assume the library already supports Python 3.10 or later. Implementing this would significantly improve the development experience, especially when working behind a company proxy.

**Comment by storkinsj** [comment]
Thank you for keeping this open. I work at one of the big four where all traffic going out is inspected. For a while azure traffic was bypassed but now we are forced to work around this.

TL;DR: Microsoft as a whole rolls out many enterprise solutions such as intune, defender, SSO which count on enterprise certificates in the MacOS keychain.  That your azure tool can not support that makes it an outlier.

I would like to add to the chorus on the "truststore" package. Per the discussion on making this the default for requests package.... I dont think that is a good idea anyway. Making the MacOS/Windows trust store "the default" for the requests package is not a good default for a system like python which is commonly used to implement linux package managers. These systems need to be independent from the normal list of default trustable entities you would find in MacOS Windows etc. 

Using "truststore" package of course does not _block_ any tool like "az" that imports "requests" to also import "truststore".  For me that tool works perfectly.

Currently, we have fallen into the practice of creating a bundle specific to the az tool and forcing it to be maintained separately from the other python related tools by wrapping your tool with a script. Before that we were going into the certifi package folder and modifying its bundle.  This does not make sense from a maintainability perspective.

additionally, exporting the whole keychain periodically (or it will break) seems like a lot to ask us to do when as I said microsoft commonly requests and/or supports requiring certificates in the system store.

Please add support for system truststore use as an option for those who are in our situation. Annoying believe me, but we have so much sensitive customer data that a breach with an RCE that attempts to egress data would be extremely impactful to any of our millions of customers. This is why we need to do traffic inspection on all egress.

**Comment by rajvcoles** [comment]
Like @storkinsj, I am also wrapping the tool with scripts, and previously, updated the certifi package bundle.

I have also tried to install pip-system-certs, which solves the problem in my environment.

```
AZ_PREFIX="$(brew --prefix azure-cli)"
AZ_PY="$AZ_PREFIX/libexec/bin/python"

"$AZ_PY" -m pip install --upgrade pip
"$AZ_PY" -m pip install pip-system-certs
```

In the current state, this problem should be prioritised from a security standpoint, as it can incentivize risky customer behavior, such as by disabling connection verification, ssl verification and other security controls, that are unrelated to the problem, to get a working state.
